pkg/apis/core/v1beta1/validation/validation_test.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+
+	"sigs.k8s.io/kubefed/pkg/apis/core/typeconfig"
+	"sigs.k8s.io/kubefed/pkg/apis/core/v1beta1"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl/enable"
+	"sigs.k8s.io/kubefed/pkg/kubefedctl/options"
+)
+
+func TestValidateFederatedTypeConfig(t *testing.T) {
+	errs := ValidateFederatedTypeConfig(validFederatedTypeConfig(), true)
+	if len(errs) != 0 {
+		t.Errorf("expected success: %v", errs)
+	}
+}
+
+func TestValidateFederatedTypeConfigSpec(t *testing.T) {
+	apiResource := &metav1.APIResource{
+		Group:      "",
+		Version:    "v1",
+		Kind:       "Service",
+		Name:       "services",
+		Namespaced: true,
+	}
+	apiResourceWithGroup := &metav1.APIResource{
+		Group:      "apps",
+		Version:    "v1",
+		Kind:       "Deployment",
+		Name:       "deployments",
+		Namespaced: true,
+	}
+
+	successCases := []*v1beta1.FederatedTypeConfig{
+		federatedTypeConfig(apiResource),
+		federatedTypeConfig(apiResourceWithGroup),
+	}
+	for _, successCase := range successCases {
+		if errs := ValidateFederatedTypeConfigSpec(&successCase.Spec, field.NewPath("spec")); len(errs) != 0 {
+			t.Errorf("expected success: %v", errs)
+		}
+	}
+
+	errorCases := map[string]*v1beta1.FederatedTypeConfig{}
+
+	// Validate required fields
```

Comment:
Please factor out the common (i.e. everything but the group) validation of the target, federated and status api resources so it can be called once for each type of resource rather than field-by-field.

Reply:
Your are correct that all 3 fields share a common struct, but that doesn't appear to me to imply that the validation paths are also shared.  

My request to clean up the testing in this method is motivated by my perception that it is not obvious that coverage is complete. There doesn't appear to be any test for the statusType field, for example, and I think the current ordering of errorCases makes it hard to easily determine whether all validation paths for a given field are being checked.

My original suggestion was to test validation of each field completely, but I take your point that it would be excessive.  What are your thoughts on the following strategy, which is already being followed in some cases?

 - test all error conditions in a method like `ValidateAPIResource` in its own test method 
   - i.e. `TestValidateAPIResource`
 - test only a single error condition for a field that is validated with a method like `ValidateAPIResource` to ensure that the validation is wired correctly

I think this would allow coverage to be observed with more clarity than what is currently proposed.

Reply:
I do think the coverage could be made a little more clear without knowledge of how the validation code paths work. I think either approach would work:
1. Testing `ValidateAPIResource` separately while also checking a single error condition for each of the FederatedTypeConfig type fields.
2. For absolute complete coverage: splitting out the common fields between TargetType, FederatedType, and StatusType into a helper test function and invoking this helper function for each of these fields.

If I'm going to revisit this in the future I think I prefer option 2.

Reply:
I think knowing how the validation paths work is preferable - these are white-box tests after all.  A correlation between the code paths and what is tested ensures that future maintainers have a reference point.